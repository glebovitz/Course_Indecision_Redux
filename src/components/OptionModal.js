import React from 'react';
import Modal from 'react-modal';

export default class OptionModal extends React.Component {
  state = {
    lastSelectedOption: ''
  };

  // Sequence: 
  //    1. props.selectedOption is cleared. 
  //    2. Modal is content updated.
  //    3. Modal then closes.
  // This causes the dialog button to jump upward.
  // Instead store selected option in a local state variabe and only change
  // when we have a new non empty selected option.
  static getDerivedStateFromProps = (nextProps, prevState) => {
    if (nextProps.selectedOption) {
        return { 
            lastSelectedOption: nextProps.selectedOption
        };
    } else {
        return null;
    }
  };

  render() {
    return (
      <Modal
        isOpen={!!this.props.selectedOption}
        onRequestClose={() => this.props.clearSelectedOption()}
        contentLabel="Selected Option"
        closeTimeoutMS={200}
        className='modal'
        ariaHideApp={false}
      >
        <h3 className='modal__title' >Select Option</h3>
        {this.state.lastSelectedOption && 
          <p className='model__body' >{this.state.lastSelectedOption}</p>}
        <button  className='button' onClick={() => this.props.clearSelectedOption()}>
          Okay
        </button>
      </Modal>
    );
  };
};
