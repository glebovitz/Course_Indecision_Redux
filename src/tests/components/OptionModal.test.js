import React from 'react';
import { shallow } from 'enzyme';
import OptionModal from '../../components/OptionModal';

test('should render OptionModal correctly', () => {
  const wrapper = shallow (
    <OptionModal
      selectedOption={undefined}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should call clearSelectedOption correctly', () => {
  const clearSelectedOption = jest.fn();
  const wrapper = shallow (
    <OptionModal
      selectedOption={undefined}
      clearSelectedOption={clearSelectedOption}
    />
  );
  wrapper.find("button").simulate("click");
  expect(clearSelectedOption).toHaveBeenCalled();
});

test('should set lastSelectedOption correctly', () => {
  const value = 'testabcde12345';

  const wrapper = shallow (
    <OptionModal
      selectedOption={value}
    />
  );
  expect(wrapper.state('lastSelectedOption')).toBe(value);
});

test('should set isOpen to false correctly', () => {
  const wrapper = shallow (
    <OptionModal
      selectedOption={undefined}
    />
  );
  expect(wrapper.find("Modal").props().isOpen).toBeFalsy();
});

test('should set isOpen to true correctly', () => {
  const wrapper = shallow (
    <OptionModal
      selectedOption={"test"}
    />
  );
  expect(wrapper.find("Modal").props().isOpen).toBe(true);
});

test('should call clearSelectedOption from onRequestClose correctly', () => {
  const clearSelectedOption = jest.fn();
  const wrapper = shallow (
    <OptionModal
      selectedOption={undefined}
      clearSelectedOption={clearSelectedOption}
    />
  );
  wrapper.find("Modal").props().onRequestClose();
  expect(clearSelectedOption).toHaveBeenCalled();
});


