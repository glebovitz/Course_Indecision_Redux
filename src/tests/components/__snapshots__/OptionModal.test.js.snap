// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render OptionModal correctly 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Selected Option"
  isOpen={false}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h3
    className="modal__title"
  >
    Select Option
  </h3>
  <button
    className="button"
    onClick={[Function]}
  >
    Okay
  </button>
</Modal>
`;
